Adds admin scripts

hlohaus opened the pull request:
### 1. Why is this change necessary?

Storefront scripts are nice, but also has many limitations.
You can't use there in a modal (iFrame) and they are not limited to admins.
With my instant tinymce app, i have a good usage for that.

### 2. What does this change do, exactly?


### 3. Describe each step to reproduce the issue or behaviour.

They are a first version of my tinymce app in the attachment.
[TinyScript.zip](https://github.com/shopware/platform/files/10569916/TinyScript.zip)

### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.


Edit: Updated example app:
[TinyScript-1.0.5.zip](https://github.com/shopware/platform/files/10612050/TinyScript-1.0.5.zip)


<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2964"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

keulinho commented:
Hey @hlohaus,

thanks for your contribution! :tada: :blue_heart: 

Can you please explain the exact use case for this change in a little bit more detail? As i have trouble getting it just from looking at the changes.

And by the way there is already the possibility to add custom routes via app scripts for the admin api, take a look at [the docs](https://developer.shopware.com/docs/guides/plugins/apps/app-scripts/custom-endpoints#admin-api-endpoints) or directly in [the code](https://github.com/shopware/platform/blob/trunk/src/Core/Framework/Script/Api/ScriptApiRoute.php#L23). 

Does that already solve your need? And if not what exactly are you missing?

I would prefer then to add the missing capabilities to the existing api scripts instead of adding a completely new hook point.
What do you think?

hlohaus commented:
@keulinho Thanks for your look. The api scripts can be used for api calls like ajax calls.
My admin script are for html pages like new tabs or iframes.
It's usefull together with the action buttons in the admin.
See screnshot: 

![tinyapp](https://user-images.githubusercontent.com/983577/216551572-e50a85d4-e898-4d3d-a1b0-2e98202079ab.jpg)
![app2](https://user-images.githubusercontent.com/983577/216552144-c30dac75-06be-4079-b3e5-8341de6987b4.jpg)

keulinho commented:
Ok i see two main issues that you are addressing here:

1. you can't return html content from custom api endpoints (that's currently limited to json data or a redirect)
2. you can't use custom api endpoints in action buttons, because the signature handling and verification

But i think your approach might not be the best to solve it, especially adding all this signature stuff to the Admin script response feels a little bit over the top IMHO.

Instead what do you think of the following ideas to solve the issues:

1. add a `raw(string $data, int $code, string $contentType)` method to the `ScriptResponseFactoryFacade`, that should address issue number one. In  your scripts you could use `include` or `macros` to split the twig scripts with the logic and those that generate the response
```
{% set content = include('includes/result.html') %}
```

2. what about to allow relative urls as well for action buttons (and probably custom-modules aka iFrames as well), because right now action button requests are proxied through the shopware server, so the admin makes a request against the shopware server, there the whole signature thingy is added to the request and then the shopware server makes a request against the action button url and validating the response etc. before it returns the result finally to the admin, in the case of using custom api routes for action buttons you don't need all the signature overhead and the extra round trip over the shopware server, in that case we can request the relative url directly and use the normal admin auth mechanism to secure the request


That are just some quick ideas, maybe i have missed something and didn't thought it through.

But all in all i think we should rethink the concept of this PR a bit, for that could you please add an `ADR` where you state exactly what problem you are facing, how you wanna solve that problem, what the implications are and especially in this case please add an example how apps may be using this, so add a example manifest and script file that shows how apps can make use of this new feature, and then let's discuss on the basis of the ADR

hlohaus commented:
Thank you very much.

To 1.: I like the raw() response idea and i found script includes:
{% set content = include('include/result.twig') %}
The directory name must be only "include".
I will change the pr.

To 2. I used the signature idea from your example app.
We need some protection for the CSRF issue.

The reasons why i created a new endpoint:
The admin api currently supports only post requests.
The admin api need a additional header for authentication.
The admin api doesn't supports `X-Frame-Options` and `CSP-Header` changes.
Maybe we find a agree here?

I will documented every new feature in the changelog:
https://github.com/hlohaus/platform/blob/admin-page/changelog/_unreleased/2023-02-02-add-admin-script.md

hlohaus commented:
Hi, the CSRF Issue are resolved by the browsers. I removed the signature service.

keulinho commented on changelog/_unreleased/2023-02-02-add-admin-script.md:
````diff
@@ -0,0 +1,52 @@
+* Bug fix: Internal resolved actions buttons now includes the complete payload. `Shopware\Core\Framework\App\ActionButton\Executor`
+* Bug fix: Use translations in your scripts. Example: 
+```
+{{ 'general.description'|trans({}, null, 'en-GB') }}
````
Nice, but i think providing a hard coded locale is a edge case, by default the correct language based on the current context should be used

hlohaus replied:
Currently i use:  {% set lang = hook.query['sw-user-language'] %}
Maybe we can fix it too.

hlohaus replied:
Should it add a setScriptsLocale method to the translator? @keulinho 
And automatic set it in a listener.

hlohaus replied:
I added a listener.

keulinho commented on src/Administration/Controller/ScriptController.php:
```diff
@@ -0,0 +1,49 @@
+        $response = $this->scriptResponseEncoder->encodeByHook($hook, $request->get('includes', []));
+
+        // Allow pages in iframes
+        $response->headers->set('X-Frame-Options', 'sameorigin');
```
IMHO this should be set by the script itself via the security service

hlohaus replied:
Good idea. I added it

keulinho commented:
Thanks for the update, but there is still no ADR and still no example of how a app will use this new features. I still don't get all the problems you wanna solve here, so please provide some more details.

I just noticed that i already built something similiar for the action buttons, refer to https://github.com/shopware/platform/commit/05626c53e1b0f6bedab8b7f6ecb899b443541ac1

hlohaus commented:
@keulinho I like your commit. Here is the example:

# Example

You can add a action button to the product details in the administration:
### manifest.xml
```
<action-button action="product-detail-tab"
  entity="product" view="detail"
  url="/api/script/open-in-new-tab">
    <label>Open editor in new window</label>
</action-button>
```
If you click on the button this app script will be executed:

### Resources/scripts/api-open-in-new-tab/index.twig
```
{% set ids = hook.request.data.ids %}
{% set entity = hook.request.data.entity %}

{% set response = services.response.json({
    "actionType": "openNewTab",
    "payload": {
        'request': hook.request,
        "redirectUrl": "/admin/app/editor?entityId=" ~ ids|first ~ "&entity=" ~ entity
    }
}) %}

{% do hook.setResponse(response) %}
```
Then you will be redirect to the page of this app script:
### Resources/scripts/admin-editor/index.twig
```
{% if hook.request %}
     // do update
     {% do services.router.redirect('editor', query) %}
{% else %}
    {% set lang = hook.query['sw-user-language'] %}
    {% set entity = services.repository.search(
        hook.query.entity, { 'ids': [hook.query.entityId]}
    ).first %}
    {% set content = include('include/editor.twig') %}
    {% set response = services.response.raw(content) %}
    {% do hook.setResponse(response) %}
{% endif %}
```
The include with the editor can be look like:
### Resources/scripts/include/editor.twig
```
<form method="post" id="content">
    <h5 class="m-3">{{ 'editor.description'|trans({}, null, lang) }}</h5>
    <textarea name="description"
                id="description">{{ entity.description }}</textarea>
</form>
```

Do you have a example for a ADR for me?

keulinho commented:
For an example you can take a look at this ADR where we introduced the custom endpoints: https://github.com/shopware/platform/blob/trunk/adr/app/2022-01-06-custom-app-api-endpoints.md

hlohaus commented:
@keulinho I added a ADR. :)

hlohaus commented:
I improved the security with a HttpOnly / SameSite Cookie.
The script url changed from /admin/app/{hook} to /api/app/{hook}

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/platform?pullRequest=2964) <br/>All committers have signed the CLA.

hlohaus committed:
Add admin scripts support with automatic authentication
Add new hook services : Router and security for admin scripts
Add integration tests for admin scripts
Refactor: Add ResponseHook interface and HookNameTrait to hooks
Add method encodeByHook in the ScriptResponseEncoder
Add RoutingExtension in the TwigEnvironment from scripts
Add translations and snippets support for admin scripts

hlohaus committed:
Fix php unit test
Fix code style in php and js

hlohaus committed:
Fix code style

hlohaus committed:
Fix unit tests

hlohaus committed:
Fix php unit and junit

shyim commented:
Hey,

We don't want to make the app scripts more complicated, and this action can be easily done with an external app-server as the action buttons and so on are async intended. Therefore, we will close. Anyway, thanks for your effort.
